File: LeetCode/DailyCodingChallenge/April 2023/Boats_to_Save_People.py
# 881. Boats to Save People
# Leetcode Medium
# Date: 3 April 2023
# TC: O(nlog(n))
# SC: O(1)
# Two Pointer Approach
from typing import List


class Solution:
    def numRescueBoats(self, people: List[int], limit: int) -> int:
        n = len(people)
        i = 0
        j = n - 1
        ans = 0
        people.sort()

        while i <= j:
            if people[i] + people[j] <= limit:
                i += 1
            j -= 1
            ans += 1

        return ans

File: Java/Array and Matrix/Boats_to_Save_People.java
// 881. Boats to Save People
// Leetcode Medium
// Date: 3 April 2023
// TC: O(nlog(n))
// SC: O(1)
// Two Pointer Approach


class Solution {

    public int numRescueBoats(int[] people, int limit) {
        int start = 0;
        int end = people.length - 1;
        int ans = 0;
        Arrays.sort(people);

        while (start <= end)
        {
            if (people[start] + people[end] <= limit)
                start += 1;
            end -= 1;
            ans += 1;
        }

        return ans;
    }
}